File: binary_tree/144.二叉树的前序遍历.go
/*
 * @lc app=leetcode.cn id=144 lang=golang
 *
 * [144] 二叉树的前序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func preorderTraversal(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	var res []int
	return preorder(root, res)
}

func preorder(root *TreeNode, res []int) []int {
	if root == nil {
		return res
	}
	res = append(res, root.Val)
	res = preorder(root.Left, res)
	res = preorder(root.Right, res)

	return res
}

// @lc code=end

File: binary_tree/144.二叉树的前序遍历.py
#
# @lc app=leetcode.cn id=144 lang=python3
#
# [144] 二叉树的前序遍历
#


# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []

        def dfs(root: Optional[TreeNode]) -> List[int]:
            if root is None:
                return res
            res.append(root.val)
            dfs(root.left)
            dfs(root.right)
            return res

        dfs(root)

        return res


# @lc code=end
